akka-sample-kafka-to-sharding-scala/processor/src/main/scala/akka/kafka/KafkaClusterSharding.scala

```diff
@@ -0,0 +1,161 @@
+   * that entities are routed to the same Entity type.
+   */
+  @ApiMayChange
+  def messageExtractor[M](system: ActorSystem,
+                          topic: String,
+                          timeout: FiniteDuration,
+                          settings: ConsumerSettings[_,_]): Future[KafkaShardingMessageExtractor[M]] =
+    getPartitionCount(system, topic, timeout, settings)
+      .map(kafkaPartitions => new KafkaShardingMessageExtractor[M](kafkaPartitions))(system.dispatcher)
+
+  /**
+   * API MAY CHANGE
+   *
+   * Asynchronously return a [[ShardingMessageExtractor]] with a default hashing strategy based on Apache Kafka's
+   * [[org.apache.kafka.clients.producer.internals.DefaultPartitioner]].
+   *
+   * The number of partitions to use with the hashing strategy will be automatically determined by querying the Kafka
+   * cluster for the number of partitions of a user provided [[topic]]. Use the [[settings]] parameter to configure
+   * the Kafka Consumer connection required to retrieve the number of partitions. Use the [[entityIdExtractor]] to pick
+   * a field from the Entity to use as the entity id for the hashing strategy.
+   *
+   * All topics used in a Consumer [[Subscription]] must contain the same number of partitions to ensure
+   * that entities are routed to the same Entity type.
+   */
+  @ApiMayChange
+  def messageExtractorNoEnvelope[M](system: ActorSystem,
+                                    topic: String,
+                                    timeout: FiniteDuration,
+                                    entityIdExtractor: M => String,
+                                    settings: ConsumerSettings[_,_]): Future[KafkaShardingNoEnvelopeExtractor[M]] =
+    getPartitionCount(system, topic, timeout, settings)
+      .map(kafkaPartitions => new KafkaShardingNoEnvelopeExtractor[M](kafkaPartitions, entityIdExtractor))(system.dispatcher)
+
+  private def getPartitionCount[M](system: ActorSystem, topic: String, timeout: FiniteDuration, settings: ConsumerSettings[_, _]): Future[Int] = {
+    implicit val ec: ExecutionContextExecutor = system.dispatcher
+    val actorNum = metadataActorCounter.getAndIncrement()
+    val consumerActor = system
+      .asInstanceOf[ExtendedActorSystem]
+      .systemActorOf(KafkaConsumerActor.props(settings), s"metadata-consumer-actor-$actorNum")
+    val metadataClient = MetadataClient.create(consumerActor, timeout)
+    val numPartitions = metadataClient.getPartitionsFor(topic).map(_.length)
+    numPartitions.map { count =>
+      system.log.info("Retrieved {} partitions for topic '{}'", count, topic)
+      count
+    }
+  }
+
+  @InternalApi
+  sealed trait KafkaClusterSharding {
+    def kafkaPartitions: Int
+    def shardId(entityId: String): String = {
+      // simplified version of Kafka's `DefaultPartitioner` implementation
+      val partition = org.apache.kafka.common.utils.Utils.toPositive(Utils.murmur2(entityId.getBytes())) % kafkaPartitions
```

Comment:
+1 to using this directly given you'll have the kafka dependency we can't add to akka